f3/api.go

```diff
@@ -46,9 +46,45 @@ type Clock interface {
 
 type Signer interface {
 	// Signs a message for the given sender ID.
-	Sign(sender ActorID, msg []byte) []byte
+	Sign(sender ActorID, msg []byte) ([]byte, error)
 	// Verifies a signature for the given sender ID.
-	Verify(sender ActorID, msg, sig []byte) bool
+	Verify(sender ActorID, msg, sig []byte) error
+}
+
+// AggSignatureScheme encapsulates signature aggregation functionality.
+// The implementation is expected to be aware of a mapping between ActorIDs and public keys.
+// A single instance of AggSignatureScheme is expected to be used for each Granite instance.
+type AggSignatureScheme interface {
+
+	// NewAggregator returns an aggregator with an "empty" aggregate signature, i.e.,
+	// one to which no signatures have yet been added.
+	// Signatures can be aggregated by calling the Add method of the returned aggregator.
+	NewAggregator() (SigAggregator, error)
+
+	// VerifyAggSig verifies the aggregate signature agg of message msg produced by a call to SigAggregator.Aggregate().
+	// Note that the SigAggregator must have been created by an AggSignatureScheme associated with the same power table
+	// as this AggSignatureScheme, otherwise the behavior of VerifyAggSig is undefined.
+	// VerifyAggSig returns nil if sig is a valid aggregate signature of msg
+	// (with respect to the associated power table) and a non-nil error otherwise.
+	VerifyAggSig(msg []byte, sig []byte) error
```

Comment:
As mentioned, I think there would be benefits from not embedding together list of signers and aggregated signature into the `sig []byte`.

Reply:
That would be very surprising! Surely a verification API must take the public keys as parameter.